test/cpp/ext/otel/otel_plugin_test.cc

```diff
@@ -1501,40 +1499,104 @@ TEST_F(OpenTelemetryPluginNPCMetricsTest, SetDoubleGauge) {
-                  kMetricName,
-                  ::testing::ElementsAre(::testing::AllOf(
-                      AttributesEq(kLabelKeys, kLabelValues, kOptionalLabelKeys,
-                                   kOptionalLabelValues),
-                      GaugeResultEq(kFinalValue),
-                      GaugeResultLaterThan(initial_timestamp))))));
+  EXPECT_THAT(
+      data,
+      ::testing::ElementsAre(::testing::Pair(
+          kMetricName,
+          ::testing::UnorderedElementsAre(
+              ::testing::AllOf(
+                  AttributesEq(kLabelKeys, kLabelValuesSet1, kOptionalLabelKeys,
+                               kOptionalLabelValuesSet1),
+                  GaugeResultIs(IntOrDoubleEq(kFinalValue)),
+                  GaugeResultLaterThan(initial_timestamp)),
+              ::testing::AllOf(
+                  AttributesEq(kLabelKeys, kLabelValuesSet2, kOptionalLabelKeys,
+                               kOptionalLabelValuesSet2),
+                  GaugeResultIs(IntOrDoubleEq(kFinalValue)))))));
+}
+
+template <typename ValueType>
+void VerifyIncrementalForSpecificMetricAndSpecificLabelSet(
+    absl::string_view metric_name,
+    absl::Span<const absl::string_view> label_key,
+    absl::Span<const absl::string_view> label_value,
+    absl::Span<const absl::string_view> optional_label_key,
+    absl::Span<const absl::string_view> optional_label_value,
+    absl::flat_hash_map<
+        std::string,
+        std::vector<opentelemetry::sdk::metrics::PointDataAttributes>>& data,
+    bool greater_than = false) {
+  std::unordered_map<std::string,
+                     opentelemetry::sdk::common::OwnedAttributeValue>
+      label_map;
+  PopulateLabelMap(label_key, label_value, &label_map);
+  PopulateLabelMap(optional_label_key, optional_label_value, &label_map);
+  ValueType prev_value = 0;
+  opentelemetry::common::SystemTimestamp prev_timestamp;
+  for (size_t i = 1; i < data[metric_name].size(); ++i) {
+    if (::testing::Matches(::testing::UnorderedElementsAreArray(
+            data[metric_name][i - 1].attributes.GetAttributes()))(label_map)) {
+      // Update the previous value for the same associated label values.
+      prev_value = opentelemetry::nostd::get<ValueType>(
+          opentelemetry::nostd::get<
+              opentelemetry::sdk::metrics::LastValuePointData>(
+              data[metric_name][i - 1].point_data)
+              .value_);
+      prev_timestamp = opentelemetry::nostd::get<
+                           opentelemetry::sdk::metrics::LastValuePointData>(
+                           data[metric_name][i - 1].point_data)
+                           .sample_ts_;
+    }
+    if (!::testing::Matches(::testing::UnorderedElementsAreArray(
+            data[metric_name][i].attributes.GetAttributes()))(label_map)) {
+      // Skip values that do not have the same associated label values.
+      continue;
+    }
+    if (greater_than) {
+      EXPECT_THAT(data[metric_name][i],
```

Comment:
Calling `EXPECT_*` in a helper function is an anti-pattern, because if this expectation fails, it will log this line number, which will make it hard to tell which invocation of this helper function is causing the problem.

The ideal approach here is to change this helper function into a matcher and use `::testing::ExplainMatchResult()` instead of `EXPECT_*`, so that ir prints the line number of the matcher invocation instead of this line number.

If that's too hard for some reason, you should at least add a `grpc_core::SourceLocation location = grpc_core::SourceLocation()` argument to this function, and then add `<< location/file)_ << ":" << location.line()` to the end of these `EXPECT_*` statements, so that it logs the line number that it was called from in addition to logging this one.